packages/embark-web3/src/index.js

```diff
@@ -17,91 +15,55 @@ class EmbarkWeb3 {
+    this.setupWeb3Api();
+    this.setupEmbarkJS();
 
-    this.events.on("blockchain:started", this.registerWeb3Object.bind(this));
-    embark.registerActionForEvent("pipeline:generateAll:before", this.addWeb3Artifact.bind(this));
     embark.registerActionForEvent("deployment:contract:deployed", this.registerInVm.bind(this));
     embark.registerActionForEvent("deployment:contract:undeployed", this.registerInVm.bind(this));
     embark.registerActionForEvent("deployment:contract:deployed", this.registerArtifact.bind(this));
     embark.registerActionForEvent("deployment:contract:undeployed", this.registerArtifact.bind(this));
+  }
+
+  async setupEmbarkJS() {
+    this.events.request("embarkjs:plugin:register", 'blockchain', 'web3', 'embarkjs-web3');
+    this.events.request("embarkjs:console:register", 'blockchain', 'web3', 'embarkjs-web3');
+    this.events.on("blockchain:started", async () => {
+      await this.registerWeb3Object()
+      this.events.request("embarkjs:console:setProvider", 'blockchain', 'web3', '{web3}');
+    });
+  }
 
-    this.registerWeb3Help();
+  async setupWeb3Api() {
+    this.events.request("runcode:whitelist", 'web3', () => { });
+    this.events.on("blockchain:started", this.registerWeb3Object.bind(this));
   }
 
   async registerWeb3Object() {
+    let checkWeb3 = `return (typeof web3 === 'undefined');`;
+    let Web3NotDefined = await this.events.request2('runcode:eval', checkWeb3);
+
+    if (!Web3NotDefined) return;
+
     const provider = await this.events.request2("blockchain:client:provider", "ethereum");
     const web3 = new Web3(provider);
     await this.events.request2("runcode:register", 'web3', web3);
     const accounts = await web3.eth.getAccounts();
     await this.events.request2('runcode:eval', `web3.eth.defaultAccount = '${accounts[0]}'`);
-  }
 
-  async registerWeb3Help() {
     await this.events.request2('console:register:helpCmd', {
       cmdName: "web3",
       cmdHelp: __("instantiated web3.js object configured to the current environment")
     });
   }
 
   async registerInVm(params, cb) {
-    let contract = params.contract;
-    let abi = JSON.stringify(contract.abiDefinition);
-    let gasLimit = 6000000;
-    const provider = await this.events.request2("blockchain:client:provider", "ethereum");
-    let contractCode = Templates.vanilla_contract({ className: contract.className, abi: abi, contract: contract, gasLimit: gasLimit, provider: provider });
-
-    try {
-      await this.events.request2('runcode:eval', contractCode);
-      let result = await this.events.request2('runcode:eval', contract.className);
-      result.currentProvider = provider;
-      await this.events.request2("runcode:register", contract.className, result);
-      cb();
-    } catch (err) {
-      cb(err);
-    }
-  }
-
-  addWeb3Artifact(_params, cb) {
-    async.map(this.config.contractsConfig.dappConnection, (conn, mapCb) => {
-      if (conn === '$EMBARK') {
```

Comment:
You lost the connection to the proxy. You need this code to change `$EMBARK` in the `dappConnection` to the proxy url.

Reply:
what do you suggest? the issue is that this refactor assumes that embarkjs itself will load blockchain.json and connect to the right thing based on the config 🤔 

what we could do is in blockchain/index.js #addArtifactFile automatically replace $Embark by the current endpoint in the generated file, what do you think?